### PR TITLE
James 1885: Test for MessageIdManager

### PR DESCRIPTION
**Pull request by quynhn**
- We do not do the contract test so we able to run it with maven
- We provide test for side effect, storing and retrieving data.

If you want new tests to be implemented, let us know

### CODE REVIEW

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        dispatcher = mock(MailboxEventDispatcher.class);
+        session = mock(MailboxSession.class);
+        quotaManager = mock(QuotaManager.class);
+
+        testingData = createTestingData(quotaManager, dispatcher);
+        messageIdManager = testingData.getMessageIdManager();
+        mailbox1 = testingData.getMailbox1();
+        mailbox2 = testingData.getMailbox2();
+
+        when(quotaManager.getMessageQuota(any(QuotaRoot.class))).thenReturn(QuotaImpl.unlimited());
```
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: You may add those behaviour in a method givenUnlimitedQuota()
And call this method in each tests, it will be easier to read.

File: protocols/imap/pom.xml [hunk 2]
```diff
@@ -122,6 +122,12 @@
             <artifactId>apache-james-mailbox-store</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (aduprat) [inline protocols/imap/pom.xml]: The TestId class may be extracted to another poject.
It's strange to depend on a mailbox project in protocols.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    @After
+    public void tearDown() {
+        testingData.clean();
+    }
+
+    @Test
+    public void deleteShouldCallEventDispatcher() throws Exception {
+        MessageId messageId = testingData.persist(mailbox1.getMailboxId(), FLAGS);
+        reset(dispatcher);
```
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: It may have the event when we persist message into "mailbox1" so it needs to reset dispatcher before another action (in this case is delete action) because we just verify event on delete function.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java (hunk 1 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: Done

File: protocols/imap/pom.xml (hunk 2 above)
Comment (quynhn) [inline protocols/imap/pom.xml]: I moved to mailbox-api project on test package and remove that dependency definition

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java [hunk 4]
```diff
@@ -68,6 +78,13 @@ public void expunged(MailboxSession session,  Map<MessageUid, MessageMetaData> u
         listener.event(eventFactory.expunged(session, uids, mailbox));
     }
 
+    public void expunged(MailboxSession session,  MessageMetaData messageMetaData, Mailbox mailbox) {
+        Map<MessageUid, MessageMetaData> metaDataMap = ImmutableMap.<MessageUid, MessageMetaData>builder()
```
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java]: It has casting error on IDE if I remove class definition

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            })
+            .toList()
+            .get(0)
+            .getUid();
+
+        assertThat(uidMessage2Mailbox1).isGreaterThan(uidMessage1Mailbox1);
+    }
+
+    @Test
+    public void setInMailboxesShouldSetHighestModSeqInNewMailbox() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: I'm not sure to understand what you test in these two tests setInMailboxesShouldSetHighestUidInNewMailbox and setInMailboxesShouldSetHighestModSeqInNewMailbox. You are using setInMailboxes, but you never check uidMessageXMailbox2. And you already have a setInMailboxesShouldNotChangeUidAndModSeqInOriginalMailbox.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    @Test
+    public void setFlagsShouldUpdateFlags() throws Exception {
+        Flags newFlags = new Flags(Flags.Flag.SEEN);
+        MessageId messageId = testingData.persist(mailbox1.getMailboxId(), FLAGS);
+
+        messageIdManager.setFlags(newFlags, MessageManager.FlagsUpdateMode.ADD, messageId, session);
+
+        MessageResult messageResult = messageIdManager.getMessages(ImmutableList.of(messageId), FetchGroupImpl.MINIMAL, session).get(0);
+        assertThat(messageResult.getFlags()).isEqualTo(newFlags);
+    }
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: Maybe you can add a test to match the contract that changing a flag change it on all its mailboxes (or not ^^).

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java [hunk 7]
```diff
@@ -56,6 +60,12 @@ public void added(MailboxSession session, SortedMap<MessageUid, MessageMetaData>
         listener.event(eventFactory.added(session, uids, mailbox));
     }
 
+    public void added(MailboxSession session, MessageMetaData messageMetaData, Mailbox mailbox) {
+        SortedMap<MessageUid, MessageMetaData> metaDataMap = new TreeMap<MessageUid, MessageMetaData>();
+        metaDataMap.put(messageMetaData.getUid(), messageMetaData);
```
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/MailboxEventDispatcher.java]: I used ImmutableSortedMap.<MessageUid, MessageMetaData>naturalOrder()

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java (hunk 3 above)
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: I'm surprised you say "may" in a test : we are supposed to know exactly what it the expected state

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    @Test
+    public void setInMailboxesShouldCallDispatcher() throws Exception {
+        MessageId messageId = testingData.persist(mailbox2.getMailboxId(), FLAGS);
+        reset(dispatcher);
+
+        messageIdManager.setInMailboxes(messageId, ImmutableList.<MailboxId>of(mailbox1.getMailboxId()), session);
+
+        verify(dispatcher).added(eq(session), any(MessageMetaData.class), any(Mailbox.class));
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: you should check the event content

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+        verifyNoMoreInteractions(dispatcher);
+    }
+
+    @Test
+    public void setInMailboxesShouldCallDispatcher() throws Exception {
+        MessageId messageId = testingData.persist(mailbox2.getMailboxId(), FLAGS);
+        reset(dispatcher);
+
+        messageIdManager.setInMailboxes(messageId, ImmutableList.<MailboxId>of(mailbox1.getMailboxId()), session);
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: you probably need a test when there's several mailboxes

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public Mailbox getMailbox2() {
+        return mailbox2;
+    }
+
+    // Should take care of find returning the MailboxMessage
+    // Should take care of findMailboxes returning the mailbox the message is in
+    // Should persist flags // Should keep track of flag state for setFlags
+    public abstract MessageId persist(MailboxId mailboxId, Flags flags);
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java]: what are these methods doing in a TestingData class ? What is the responsibility of the class ?

File: protocols/imap/pom.xml (hunk 2 above)
Comment (mbaechler) [inline protocols/imap/pom.xml]: yes, that's the right solution IMO

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java (hunk 10 above)
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java]: I guess your reproaches are based on a naming issue.

I propose : MessageIdManagerSystem

(Pros : Consistent with MPT)

I also like the term "Pilot" as I use this class to Pilot the content of the DB for MessageIdManager... Let say it do not look conventionnal, so let's skip it.

**Comment by chibenwa** [comment]
I'll fixup this before leaving the office, as there's no serious stuff.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        return mailbox1;
+    }
+
+    public Mailbox getMailbox2() {
+        return mailbox2;
+    }
+
+    // Should take care of find returning the MailboxMessage
+    // Should take care of findMailboxes returning the mailbox the message is in
+    // Should persist flags // Should keep track of flag state for setFlags
```
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MessageIdManagerTestingData.java]: Done

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java (hunk 5 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: It have some assert issues and I changed it.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java (hunk 6 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: I think the flag is related to messages, it does not related to mailbox so we don't need to check message's flag on mailbox. Is that correct?

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: Before MessageIdManager, a message is always a MailboxMessage, so it is related to a mailbox.
Introducing MessageIdManager we can determinate messages outside Mailboxes, so we need to define a behaviour as at the storage level Messages could be duplicated between Mailboxes.

File: protocols/pom.xml [hunk 12]
```diff
@@ -243,6 +243,13 @@
                 <version>${project.version}</version>
             </dependency>
             <dependency>
+                <groupId>org.apache.james</groupId>
+                <artifactId>apache-james-mailbox-store</artifactId>
+                <scope>test</scope>
+                <type>test-jar</type>
+                <version>${project.version}</version>
+            </dependency>
+            <dependency>
```
Comment (rouazana) [inline protocols/pom.xml]: protocols should not depend on store.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            .get(0);
+        SimpleMessageMetaData metadataMailbox3 = fromMessageResult(messageId, messageResultMaibox3);
+
+        verify(dispatcher).added(session, metadataMailbox1, mailbox1);
+        verify(dispatcher).added(session, metadataMailbox3, mailbox3);
+        verifyNoMoreInteractions(dispatcher);
+    }
+
+    @Test
+    public void setInMailboxesThrowWhenOverQuota() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: missing "Should"

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java [hunk 14]
```diff
@@ -32,7 +32,7 @@
 
 public interface MessageIdManager {
 
-    void setFlags(Flags newState, FlagsUpdateMode replace, MessageId messageId, MailboxId mailboxId, MailboxSession mailboxSession) throws MailboxException;
+    void setFlags(Flags newState, FlagsUpdateMode replace, MessageId messageId, MailboxSession mailboxSession) throws MailboxException;
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageIdManager.java]: So can you add back the MailboxId parameter (or MailboxIds if you want) to be more consistent with MessageManager API and IMAP?

**Comment by rouazana** [comment]
The 2 last commits should be fixups, no?

**Comment by quynhn** [comment]
```
The 2 last commits should be fixups, no?
```
Yes, I break it and changed the commit messages

**Comment by chibenwa** [comment]
Can we have some more tests?
 
 - side effect : I set flags in a mailbox the message is in. Dispatcher should not be called.
 - storage :  I set flags in a mailbox the message is in. The message keeps its flags.

**Comment by rouazana** [comment]
yes, please add some more tests as asked by benoit + empty list test and list with 2 elements tests (matching/non matching/etc.)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 15]
```diff
@@ -310,6 +322,60 @@ public void setFlagsShouldNotChangeFlagsOfAnotherMessageInSameMailbox() throws E
...
+        messageIdManager.setFlags(newFlags, MessageManager.FlagsUpdateMode.ADD, messageId1, ImmutableList.of(mailbox1.getMailboxId(), mailbox2.getMailboxId()), session);
+
+        final List<Map.Entry<MessageId, Flags>> entries = FluentIterable.from(messageIdManager.getMessages(ImmutableList.of(messageId1, messageId2), FetchGroupImpl.MINIMAL, session))
+            .transform(toMapEntryOfFlags())
+            .toList();
+        ImmutableMap.Builder<MessageId, Flags> builder = ImmutableMap.builder();
+        for (Map.Entry<MessageId, Flags> entry : entries) {
+            builder.put(entry);
+        }
+        Map<MessageId, Flags> flags = builder.build();
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: ImmutableMap.builder().putAll(entries).build() ?

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: or even ImmutableMap.copyOf(entries)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 16]
```diff
@@ -310,6 +322,60 @@ public void setFlagsShouldNotChangeFlagsOfAnotherMessageInSameMailbox() throws E
...
+
+        assertThat(flags).hasSize(2);
+        assertThat(flags.get(messageId1)).isEqualTo(newFlags);
+        assertThat(flags.get(messageId2)).isEqualTo(FLAGS);
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: this one should not be changed?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java (hunk 15 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: I tried but all of those don't work on the IDE. Those require Map but we have Map.Entry

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: ok, seems to be in an other version of Guava. And what about using directly FluentIterable.toMap?

Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: I tried with toMap but it will build the Map with key is MessageResult (Object in list) but here we need MessageId (Property of object in list)

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: ok

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
test this please

**Comment by chibenwa** [comment]
We miss tests for setInMailboxes removing messages from mailboxes : 
 - event dispatching
 - effects on storage

**Comment by rouazana** [comment]
test this please
